src/main/java/adris/altoclef/tasks/misc/GetToOuterEndIslandsTask.java

```diff
@@ -0,0 +1,120 @@
+            new Vec3i(2, -1, 0),
+            new Vec3i(0, -1, 2),
+            new Vec3i(-2, -1, 0),
+            new Vec3i(0, -1, -2)
+    };
+    public GetToOuterEndIslandsTask() {
+
+    }
+
+    @Override
+    protected void onStart(AltoClef mod) {
+        mod.getBehaviour().push();
+        mod.getBlockTracker().trackBlock(Blocks.END_GATEWAY);
+        _beatTheGame = new BeatMinecraft2Task();
+    }
+
+    @Override
+    protected Task onTick(AltoClef mod) {
+        if (isFinished(mod)) {
+            setDebugState("What?");
+            return null;
+        }
+        if (mod.getBlockTracker().anyFound(Blocks.END_GATEWAY)) {
+            if (!mod.getItemStorage().hasItemInventoryOnly(Items.ENDER_PEARL)) {
+                setDebugState("Getting an ender pearl");
+                return new CataloguedResourceTask(new ItemTarget(Items.ENDER_PEARL, 1));
+            }
+            BlockPos gateway = mod.getBlockTracker().getNearestTracking(Blocks.END_GATEWAY).get();
+            int blocksNeeded = Math.abs(mod.getPlayer().getBlockY() - gateway.getY()) +
+                    Math.abs(mod.getPlayer().getBlockX() - gateway.getX()) +
+                    Math.abs(mod.getPlayer().getBlockZ() - gateway.getZ()) - 1;
+            if (StorageHelper.getBuildingMaterialCount(mod) < blocksNeeded) {
+                setDebugState("Getting building materials");
+                return new GetBuildingMaterialsTask(blocksNeeded);
+            }
+            GoalAnd goal = makeGoal(gateway);
+            Debug.logMessage(mod.getPlayer().getBlockPos().toString());
+            if (!goal.isInGoal(mod.getPlayer().getBlockPos()) || !mod.getPlayer().isOnGround()) {
+                mod.getClientBaritone().getCustomGoalProcess().setGoal(goal);
+                if (!mod.getClientBaritone().getPathingBehavior().isPathing()) {
+                    mod.getClientBaritone().getCustomGoalProcess().path();
+                }
+                setDebugState("Getting close to gateway...");
+                return null;
+            }
+            setDebugState("Throwing the pearl inside");
+            return new InteractWithBlockTask(Items.ENDER_PEARL, gateway);
+        }
+        setDebugState("Beating the Game to get to an end gateway");
+        return _beatTheGame;
+    }
+
+    @Override
+    protected void onStop(AltoClef mod, Task interruptTask) {
+        mod.getBlockTracker().stopTracking(Blocks.END_GATEWAY);
+        mod.getBehaviour().pop();
+    }
+
+    @Override
+    protected boolean isEqual(Task other) {
+        return other instanceof GetToOuterEndIslandsTask;
+    }
+
+    @Override
+    public boolean isFinished(AltoClef mod) {
+        return WorldHelper.getCurrentDimension() == Dimension.END &&
+                mod.getPlayer().getPos().distanceTo(new Vec3d(0, 64, 0)) > 400;
+    }
+
+    @Override
+    protected String toDebugString() {
+        return "Going to outer end islands";
+    }
+
+    private GoalAnd makeGoal(BlockPos gateway) {
```

Comment:
I don't understand the need for the offsets. IIRC, you mentioned this is to avoid getting stuck and/or hitting the bedrock. That sounds like a pearl recalculation to me. This may work, but it would be a good idea to think about why the first attempt failed. Sometimes duct tape solutions can cause further problems down the road. Since this works, take this as a suggestion.

Reply:
Aight